File: kotlin/src/main/kotlin/673.NumberofLongestIncreasingSubsequence.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/number-of-longest-increasing-subsequence/


class NumberofLongestIncreasingSubsequence{
  class Node(val begin: Int, val end: Int){
    val md = begin + (end - begin)/2
    val left by lazy { Node(begin, md) }
    val right by lazy { Node(md+1, end) }
    var cur = 0 to 1

    fun merge(lenCount1: Pair<Int, Int>, lenCount2: Pair<Int, Int>): Pair<Int, Int>{
      val (l1, c1) = lenCount1
      val (l2, c2) = lenCount2
      if(l1 == l2) return l1 to (if(l1==0) 1 else c1+c2)
      return if(l1 < l2) lenCount2 else lenCount1
    }

    fun insert(num: Int, lenCount: Pair<Int, Int>){
      if(num < begin || end < num) return
      cur = if(begin == end)
        merge(cur, lenCount)
      else{
        if(num <= md) left.insert(num, lenCount)
        else right.insert(num, lenCount)
        merge(left.cur, right.cur)
      }
    }

    fun query(num: Int): Pair<Int, Int>{
      if(end <= num || begin == end) return cur
      if(num <= md) return left.query(num)
      return merge(left.cur, right.query(num))
    }
  }

  fun findNumberOfLIS(nums: IntArray): Int {
    val n = nums.size
    if(n == 0) return 0
    val min = nums.min()!!
    val root = Node(min, nums.max()!!)
    for(num in nums){
      val (len, count) = if(num == min) 0 to 1 else root.query(num-1)
      root.insert(num,len+1 to count)
    }
    return root.cur.second
  }

  @Test
  fun test0(){
    assertEquals(4, findNumberOfLIS(intArrayOf(4,3,2,1)))
  }

  @Test
  fun test1(){
    assertEquals(2, findNumberOfLIS(intArrayOf(1,3,5,4,7)))
  }

  @Test
  fun test2(){
    assertEquals(5, findNumberOfLIS(intArrayOf(2,2,2,2,2)))
  }
}

File: python/673_NumberofLongestIncreasingSubsequence.py
from unittest import TestCase
# https://leetcode.com/problems/number-of-longest-increasing-subsequence/
import bisect


class NumberofLongestIncreasingSubsequence(TestCase):
    def findNumberOfLIS(self, nums: 'List[int]') -> 'int':
        n = len(nums)
        ends = [[] for _ in range(n)]
        for num in nums:
            l, r = 0, n-1
            while l <= r:
                m = (l+r) >> 1
                if ends[m] and ends[m][-1][0] < num:
                    l = m+1
                else:
                    r = m-1
            if l-1 >= 0:
                cnt = 0
                for v, c in reversed(ends[l-1]):
                    if v >= num: break
                    cnt += c
            else:
                cnt = 1
            ends[l].append((num, cnt))
        for ed in reversed(ends):
            if ed:
                return sum(map(lambda e: e[1], ed))
        return 0

    # concise
    def findNumberOfLIS(self, nums: 'List[int]') -> 'int':
        n, MAX = len(nums), float("inf")
        ends = [[MAX, []] for _ in range(n)]
        for v in nums:
            i = bisect.bisect_left(ends, [v, []])
            cnt = sum(c for x, c in ends[i-1][1] if x < v) if i else 1
            ends[i][0] = v
            ends[i][1].append((v, cnt))
        return sum(c for v, c in next((cnts for mv, cnts in reversed(ends) if mv != MAX), []))

    def test0(self):
        self.assertEqual(2, self.findNumberOfLIS([1,8,9,7,6,3,4,2,1,7,8,6,5,11,10]))

    def test1(self):
        self.assertEqual(2, self.findNumberOfLIS([1,3,5,4,7]))

    def test2(self):
        self.assertEqual(5, self.findNumberOfLIS([2,2,2,2,2]))
